File: kt/BeginnersSelection/10_Traveling.kt
// https://atcoder.jp/contests/abs/tasks/arc089_a

import kotlin.math.abs

fun main(args: Array<String>) = (
  readLine()!!
    .toInt()
    .let { n ->
      (1..n).map {
        readLine()!!
          .split(" ")
          .map(String::toInt)
          .let { x -> Triple(x[0], x[1], x[2])}
      }
      .fold(Triple(0, 0, 0)) { a: Triple<Int, Int, Int>?, x ->
        a?.let {
          if (((x.first and 1) == (x.second + x.third and 1))
            && ((x.first - a.first) >= (abs(x.second - a.second) + abs(x.third - a.third)))
          )
            x
          else
            null
        } 
      }
    }
    .let { it?.let { "Yes" } ?: "No" }
    .let(::println)
)

// https://atcoder.jp/contests/abs/submissions/21020882

File: scala/BeginnersSelection/10_Traveling.scala
// https://atcoder.jp/contests/abs/tasks/arc089_a

object Main {

  implicit class Pipeline[T](x: T) {
    def |>[S](f: T => S): S = f(x)
  }

  def main(args: Array[String]) = (
    io.StdIn.readLine.toInt
    |> ((n) =>
      (1 to n).map(_ => (
        io.StdIn.readLine
          .split(" ")
          .map(_.toInt)
          |> ((x) => (x(0), x(1), x(2)))
      ))
      .foldLeft(Option(0, 0, 0)) { case (acc, (t2, x2, y2)) =>
        acc match {
          case Some((t1, x1, y1))
            if (((t2 & 1) == (x2 + y2 & 1))
              && ((t2 - t1) >= (math.abs(x2 - x1) + math.abs(y2 - y1)))
            ) => Some((t2, x2, y2))
          case _ => None
        }
      }
    )
    |> ((x) =>
      x match {
        case Some(x) => "Yes"
        case _       => "No"
      }
    )
    |> println
  )
}

// https://atcoder.jp/contests/abs/submissions/21020571
